wolff/optimization/gradient_descent_solver.py

```diff
@@ -164,6 +177,19 @@ def simulate(self):
         self._model.theta = self._theta
         self._model.process(self.grid)
 
+        self._error = LA.norm(self._model.J)
+
+        self.Print()
+
+        with open(self._output_file, mode='w') as initial:
```

Comment:
Should use `self.output_file`
```suggestion
        with open(self.output_file, mode='w') as initial:
```

Reply:
Do you allow no log file? `if self.output_file is not None:`